pageserver/src/tenant/storage_layer/inmemory_layer.rs

```diff
@@ -371,26 +371,23 @@ impl InMemoryLayer {
 
         let mut planned_block_reads = BinaryHeap::new();
 
-        for range in keyspace.ranges.iter() {
-            let mut key = range.start;
-            while key < range.end {
-                if let Some(vec_map) = inner.index.get(&key) {
-                    let lsn_range = match reconstruct_state.get_cached_lsn(&key) {
-                        Some(cached_lsn) => (cached_lsn + 1)..end_lsn,
-                        None => self.start_lsn..end_lsn,
-                    };
-
-                    let slice = vec_map.slice_range(lsn_range);
-                    for (entry_lsn, pos) in slice.iter().rev() {
-                        planned_block_reads.push(BlockRead {
-                            key,
-                            lsn: *entry_lsn,
-                            block_offset: *pos,
-                        });
-                    }
-                }
+        for (key, vec_map) in inner.index.iter() {
+            let key = key.clone();
+            if !keyspace.overlaps(&(key.clone()..key.next())) {
+                continue;
+            }
+            let lsn_range = match reconstruct_state.get_cached_lsn(&key) {
+                Some(cached_lsn) => (cached_lsn + 1)..end_lsn,
+                None => self.start_lsn..end_lsn,
+            };
```

Comment:
I wrote this comment before I had fully understood what you were trying to do :)

> Given that we have range1, range2, range3, we can simply create a B+ tree iterator on these 3 ranges directly, instead of getting key one by one.

I agree that we need a better solution here, but I don't follow your proposal.